common/src/main/java/com/wynntils/core/features/overlays/OverlayManager.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright © Wynntils 2022.
+ * This file is released under AGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.core.features.overlays;
+
+import com.wynntils.core.WynntilsMod;
+import com.wynntils.core.features.overlays.annotations.OverlayInfo;
+import com.wynntils.mc.event.RenderEvent;
+import com.wynntils.mc.utils.McUtils;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+
+public class OverlayManager {
+    private static final Map<Overlay, OverlayInfo> overlayInfoMap = new HashMap<>();
+
+    private static final Set<Overlay> enabledOverlays = new HashSet<>();
+
+    public static void registerOverlay(Overlay overlay, OverlayInfo overlayInfo) {
+        overlayInfoMap.put(overlay, overlayInfo);
+
+        if (overlayInfo.enabled()) {
+            enabledOverlays.add(overlay);
+        }
+    }
+
+    public static void disableAllOverlaysForFeature(List<Overlay> overlays) {
+        enabledOverlays.removeIf(overlays::contains);
+    }
+
+    public static void enableAllOverlaysForFeature(List<Overlay> overlays) {
+        enabledOverlays.addAll(overlays);
+    }
+
+    @SubscribeEvent
+    public static void onRenderPre(RenderEvent.Pre event) {
+        McUtils.mc().getProfiler().push("preRenOverlay");
+        renderOverlays(event, OverlayInfo.RenderState.Pre);
+        McUtils.mc().getProfiler().pop();
+    }
+
+    @SubscribeEvent
+    public static void onRenderPost(RenderEvent.Post event) {
+        McUtils.mc().getProfiler().push("postRenOverlay");
+        renderOverlays(event, OverlayInfo.RenderState.Post);
+        McUtils.mc().getProfiler().pop();
+    }
+
+    private static void renderOverlays(RenderEvent event, OverlayInfo.RenderState renderState) {
+        for (Overlay overlay : enabledOverlays) {
+            OverlayInfo annotation = overlayInfoMap.get(overlay);
+
+            if (renderState != annotation.renderAt() || event.getType() != annotation.renderType()) continue;
+
+            overlay.render(overlay.getPosition(), event.getPoseStack(), event.getPartialTicks(), event.getWindow());
```

Comment:
When I see this code, I'm starting to think that a perhaps even simpler approach would be to store an `boolean enabled` in the Overlay class, and turn it on and off when enabling/disabling, and just skip the overlay in this for loop if it's not enabled. Otoh, it does not matter that much, and the current implementation with `enabledOverlays` is fine. Change it if you think it would be a nicer pattern, to store the knowledge about denabled/disabled in the Overlay itself, otherwise keep it if you think this is a responsibility of the Manager to know about.

Reply:
Easier to get all enabled Overlays with this approach. Also with this style, we follow the "overlay doesn't have a understanding of state, and is only an abstraction of rendering on screen with specific settings", which is how I  think about overlays.

Reply:
I accept this. But with just a clarification: I think overlays can have some understanding of state, but that relates to "model state" (like the set of messages to show in the message pane). It does now know if it is shown or not, it just renders itself if asked to.